### PR TITLE
Omega-0 Design Document: Requirements and Testing only

### PR DESCRIPTION
**Pull request by mark-petersen**
This PR adds the design document for Omega-0, which will solve the shallow water equations with identical vertical layers and inactive tracers. It only includes the sections on Overview, Requirements, and Testing. Sections on Algorithmic Formulation and Design will be added in a subsequent PR.

### CODE REVIEW

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+A substantial investment has already been made in the polaris tools. Continuing the use of polaris for Omega will speed development and encourage the documentation and long-term reproducibility of test cases.
+
+The test cases relevant to this design document are in Section 5 below.
+
+### 2.5 Requirement: Omega-0 will run portably on various DOE architectures (CPU and GPU nodes)
+
+Omega will be able to run on all the upcoming DOE architectures and make good use of GPU hardware. This should occur with minimal alterations in the high-level PDE solver code for different platforms. 
+
+The proposed solution is to use YAKL [(Norman et al. 2023)](https://link.springer.com/10.1007/s10766-022-00739-0) for within-node, shared memory computations. All large arrays will be YAKL objects, and computations that loop over array indices will use YAKL function calls.
```
Comment (xylar) [inline components/omega/doc/design/ShallowWaterOmega0.md:69]: Pedantic but maybe this should be in Design, no in requirements?

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+### 2.9 Requirement: Performance will be at least as good or better than MPAS-Ocean
+
+Performance will assessed with the following metrics:
+
+1. Single CPU throughput
+1. Parallel CPU scalability to high node counts
+1. Single GPU throughput
+
+The first two items will be compared to MPAS-Ocean with the same test configuration.
```
Comment (xylar) [inline components/omega/doc/design/ShallowWaterOmega0.md:91]: Is it worth describing what we mean by MPAS-Ocean in the same configuration, E.g. what terms, parameterizations, etc. are on and what are turned off?

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+- automation and reproducibility in polaris
+
+### 5.6 Demonstration cases with realistic coastlines
+
+Tests in realistic domains will facilitate the testing of boundary conditions on realistic coastlines, which are not included in any of the previous tests. They will also be useful for highlights and talks to show that Omega-0 can run in global realistic domains, albeit with shallow water equations only.
+
+Potential test cases include:
+
+1. Coastal inundation: Delaware Bay mesh, after ICOM tests, to demonstrate boundary conditions, resolution, and for 'splashy' images. Would tidal forcing be required for that?
+1. Global realistic Tsunami to test boundary conditions and grid-scale noise. See Darren Engwirda's recent tests on Alaska's Aleutian Islands.
```
Comment (vanroekel) [inline components/omega/doc/design/ShallowWaterOmega0.md:203]: Don't (1) and (2) require the unified earth system mesh + wetting and drying from ICoM?  I'm pretty concerned about listing these as possible tests.

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+Requirements for tests are:
+
+- Single CPU performance should be as good or better than single CPU MPAS-Ocean.
+- GPU tests on single node should be better than single node CPU tests.
+- Scaling on CPUs multi-node, up to 4096 cores or higher, should be close to perfect scaling for 512x512 mesh, and as good or better than MPAS-Ocean.
+- Scaling on GPUs for multi-node should be close to perfect scaling for 512x512 mesh.
+- automation and reproducibility in polaris
+
+### 5.6 Demonstration cases with realistic coastlines
```
[the remark this answers is not in the document]
Comment (cbegeman) [reply components/omega/doc/design/ShallowWaterOmega0.md:196]: We don't yet have a good, simple, barotropic test case that feels suitable for coastlines. Any other suggestions there? I'm not convinced that the Stommel gyre case is similar enough to coastlines, having only non-periodic planar boundaries. What do you (@vanroekel @sbrus89 @mark-petersen) think of putting this off until Omega-1? At that point we could do the SOMA test case, which had curved boundaries more like a basin with coastlines.

[the remark this answers is not in the document]
Comment (vanroekel) [reply components/omega/doc/design/ShallowWaterOmega0.md:196]: @cbegeman waiting until omega-1 seems like a very sensible idea to me.  I also fully agree with you that the idealized gyre cases with vertical walls are not super similar to coastlines

File: components/omega/doc/design/ShallowWaterOmega0.md (hunk 1 above)
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md]: Good point. I changed this to a list of possible options to provide some background, but will not specify the solution until the design section.

File: components/omega/doc/design/ShallowWaterOmega0.md (hunk 2 above)
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md]: Thanks. I added those details. Conveniently, this will be published soon in the Bishnu GMD paper.

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+This test should be conducted as soon as momentum and thickness equations and time-stepping is in place. It does not require any tracer infrastructure.
+
+### 5.4 Advection of cosine bell on a sphere
+
+Advection of a tracer cosine bell around the sphere was described in [Williamson et al. 1992](https://www.sciencedirect.com/science/article/pii/S0021999105800166) as test case 1. It tests the tracer advection and tracer time stepping. Velocity and thickness are predefined fields and remain fixed, so this does not exercise those modules.
+
+Requirements for tests are:
+
+- expected order of convergence (check)
```
[the remark this answers is not in the document]
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md]: done, thanks.

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+Question: should we make tests for spherical domain here as well? Hyun has done those before with spherical harmonics.
+
+### 5.2 Inertia Gravity Wave: linearized shallow water, no tracers
+
+The inertia gravity wave test provides an exact solution in time for the linearized shallow water equations (momentum and thickness). It tests the time stepping scheme along with the linearized advection terms, the SSH gradient. It does not test diffusive terms or bottom drag. It is conducted on a doubly periodic cartesian mesh, so does not test boundary conditions. Error should converge at 1st order, as described by [Bishnu 2021](https://zenodo.org/record/7439539) and Bishnu et al. 2023. Also see the inertia gravity test case in Polaris.
+
+Requirements for tests are:
+
+- expected order of convergence
```
[the remark this answers is not in the document]
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md:159]: done, thanks.

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+- conservation of total volume
+- automation and reproducibility in polaris
+
+This test should be conducted as soon as momentum and thickness equations and time-stepping is in place. It does not require any tracer infrastructure.
+
+### 5.4 Advection of cosine bell on a sphere
```
[the remark this answers is not in the document]
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md]: Thanks, this is better wording.

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+Advection of a tracer cosine bell around the sphere was described in [Williamson et al. 1992](https://www.sciencedirect.com/science/article/pii/S0021999105800166) as test case 1. It tests the tracer advection and tracer time stepping. Velocity and thickness are predefined fields and remain fixed, so this does not exercise those modules.
```
[the remark this answers is not in the document]
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md]: done.

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+Requirements for tests are:
+
+- expected order of convergence (check)
+- conservation of total tracer amount
```
[the remark this answers is not in the document]
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md:185]: done

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+- tracer min and max values should remain bounded by initial bounds.
+- automation and reproducibility in polaris
+
+### 5.5 Performance testing
+
+Tests can be conducted with inertia-gravity wave tests but with full non-linear terms and 100 identical layers. Domain will be cartesian resolutions from 64x64 up to 512x512. These can be compared with Bishnu 2023 (ref). Additionally, we could test spherical cases with a Williamson test case or similar.
```
[the remark this answers is not in the document]
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md]: I changed the wording to be more specific. One of the Bishnu papers has all these performance comparisions with protocol and results.

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+Note: consider applying this as a unit test.
+
+Note: We may not need this group of tests long-term, because the inertia gravity wave tests all of them together.
+
+Question: should we make tests for spherical domain here as well? Hyun has done those before with spherical harmonics.
+
+### 5.2 Inertia Gravity Wave: linearized shallow water, no tracers
+
+The inertia gravity wave test provides an exact solution in time for the linearized shallow water equations (momentum and thickness). It tests the time stepping scheme along with the linearized advection terms, the SSH gradient. It does not test diffusive terms or bottom drag. It is conducted on a doubly periodic cartesian mesh, so does not test boundary conditions. Error should converge at 1st order, as described by [Bishnu 2021](https://zenodo.org/record/7439539) and Bishnu et al. 2023. Also see the inertia gravity test case in Polaris.
```
[the remark this answers is not in the document]
Comment (mark-petersen) [reply components/omega/doc/design/ShallowWaterOmega0.md]: Yes, thanks. Changed. I can even point to the specific plot in Sids paper when it is published (soon)!

File: components/omega/doc/design/ShallowWaterOmega0.md [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+For these comparisons, MPAS-Ocean will be set up with the identical terms and functionality as Omega-0. This means that vertical advection, vertical mixing, and all parameterizations will be disabled. Comparisons with these terms will be made with Omega-1 and higher and will be described in future design documents.
+
+### 2.10 Requirement: Full-node GPU throughput will be comparable or better than full-node CPU throughput
+
+For GPU throughput, comparisons should be made between full-node CPU throughput and full-node GPU throughput. For example, [Perlmutter at NERSC](https://docs.nersc.gov/systems/perlmutter/architecture/) has nodes with 64 and 128 CPU cores (AMD EPYC 7763) and 4 GPUs (NVIDIA A100). We expect the full-node GPU throughput to be at least as good as the full-node CPU throughput, and potentially a factor of four higher. These numbers depend on the performance specifications of the particular hardware.
```
Comment (mark-petersen) [inline components/omega/doc/design/ShallowWaterOmega0.md:118]: @philipwjones and @sarats, is this a reasonable requirement for GPU performance for Omega-0? I'm happy to alter or reword.

Comment (philipwjones) [reply components/omega/doc/design/ShallowWaterOmega0.md:118]: It's fine for a requirement. Aspirations are for higher, but that's mentioned too.

**Comment by mark-petersen** [comment]
I added everyone who commented on as reviewers. If you are happy with my changes, please approve this week so I can merge.

**Review by cbegeman — APPROVED** [review]
@mark-petersen Thanks for your work on this. The latest changes look good to me!

**Review by philipwjones — APPROVED** [review]
Looks fine for this stage.

**Review by sbrus89 — APPROVED** [review]
Looks good to me. Thanks @mark-petersen!

**Review by vanroekel — APPROVED** [review]
looks good to me.
